Source/UI/NewUI/BillingForm/ViewModel/DefaultBillingFormViewModel.swift

```diff
@@ -0,0 +1,156 @@
+        validate(text: textField.text , cellStyle: textField.type, row: textField.tag)
+        
+        if !(textField.text?.isEmpty ?? true) || textField.type.style?.isOptinal ?? false {
+            textValueOfCellType[textField.type.hash] = textField.text
+        } else {
+            textValueOfCellType[textField.type.hash] = nil
+            
+        }
+        updatedRow = textField.tag
+    }
+    
+    func textFieldShouldChangeCharactersIn(textField: UITextField, replacementString string: String) {
+        guard let textField = textField as? BillingFormTextField else { return }
+        validate(text: string , cellStyle: textField.type, row: textField.tag)
+        
+        if !(string.isEmpty)  {
+            textValueOfCellType[textField.type.hash] = string
+        } else if textField.text?.count ?? 1 == 1, !(textField.type.style?.isOptinal ?? false)  {
+            textValueOfCellType[textField.type.hash] = nil
+        }
+        
+        let isSuccessful =  textValueOfCellType.values.count == style.cells.count && !(errorFlagOfCellType.isEmpty || errorFlagOfCellType.values.allSatisfy({$0}))
+        editDelegate?.didFinishEditingBillingForm(successfully:  isSuccessful)
+    }
+    
+    
+    func tableView(estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
+        UITableView.automaticDimension
+    }
+    
+    func tableView(numberOfRowsInSection section: Int) -> Int {
+        style.cells.count
+    }
+    
+    func tableView(tableView: UITableView, cellForRowAt indexPath: IndexPath, sender: UIViewController) -> UITableViewCell {
+        getCell(row: indexPath.row, delegate: sender as? BillingFormViewController)
+    }
+    
+    func doneButtonIsPressed(sender: UIViewController) {
+        
+        let phone = Phone(
+            number: textValueOfCellType[BillingFormCell.phoneNumber(nil).hash],
+            country: nil)
+        
+        let address = Address(
+            addressLine1: textValueOfCellType[BillingFormCell.addressLine1(nil).hash],
+            addressLine2: textValueOfCellType[BillingFormCell.addressLine2(nil).hash],
+            city: textValueOfCellType[BillingFormCell.city(nil).hash],
+            state: textValueOfCellType[BillingFormCell.state(nil).hash],
+            zip: textValueOfCellType[BillingFormCell.postcode(nil).hash],
+            country: nil)
+                                                       
+                                                      
+        delegate?.onTapDoneButton(address: address, phone: phone)
+        sender.dismiss(animated: true)
+    }
+    
+    func cancelButtonIsPressed(sender: UIViewController) {
+        sender.dismiss(animated: true)
+    }
+    
+    internal func validate(text: String?, cellStyle: BillingFormCell, row: Int)  {
+        guard let style = cellStyle.style, !style.isOptinal else {
+            errorFlagOfCellType[cellStyle.hash] = false
```

Comment:
are we sure cellStyle.hash is always within bounds of errorFlagOfCellType, do we need have a check something like below 
`if cellStyle.hash <=  errorFlagOfCellType.count`